ql/src/java/org/apache/hadoop/hive/ql/udf/ptf/ValueBoundaryScanner.java

```diff
@@ -664,6 +632,87 @@ public Object computeValue(Object row) throws HiveException {
    */
   public abstract boolean isEqual(Object v1, Object v2);
 
+  protected int binarySearchBack(int r, PTFPartition p, Object sortKey, int amt,
+      Order order) throws HiveException {
+    boolean isOrderDesc = order.equals(Order.DESC);
+    Object rowVal = sortKey;
+
+    int rMin = -1;  // tracks lowest possible number fulfilling the range requirement
+    int rMax = r; // tracks highest possible number fulfilling the range requirement
+
+    boolean isDistanceGreater = true;
+    while (true) {
+      if (rMin == rMax) {
+        return rMin;
+      }
+      isDistanceGreater =
+          isDistanceGreater(isOrderDesc ? rowVal : sortKey, isOrderDesc ? sortKey : rowVal, amt);
+      if (!isDistanceGreater && r == 0) {
```

Comment:
we need this, otherwise there is a scenario, when the algorithm finds that we need to step further, even beyond 0, so:
```
isDistanceGreater = false
r == 0
```
in this case rMax becomes rowId (which is 0), and as rMin is still -1, so:
```
rowId = rMin + (rMax - rMin) / 2 = -1
```

which could lead to this exception:
```
java.lang.IndexOutOfBoundsException: Index: -1, Size: 21
	at java.util.LinkedList.checkElementIndex(LinkedList.java:555)
	at java.util.LinkedList.get(LinkedList.java:476)
	at org.apache.hadoop.hive.ql.udf.ptf.TestBoundaryCache.lambda$setupMocks$3(TestBoundaryCache.java:256)
	at org.apache.hadoop.hive.ql.exec.PTFPartition.getAt(PTFPartition.java:120)
	at org.apache.hadoop.hive.ql.udf.ptf.ValueBoundaryScanner.computeValueUseCache(ValueBoundaryScanner.java:302)
	at org.apache.hadoop.hive.ql.udf.ptf.SingleValueBoundaryScanner.binarySearchBack(ValueBoundaryScanner.java:657)
	at org.apache.hadoop.hive.ql.udf.ptf.SingleValueBoundaryScanner.binarySearchBack(ValueBoundaryScanner.java:635)
	at org.apache.hadoop.hive.ql.udf.ptf.SingleValueBoundaryScanner.computeStartPreceding(ValueBoundaryScanner.java:423)
	at org.apache.hadoop.hive.ql.udf.ptf.SingleValueBoundaryScanner.computeStartPreceding(ValueBoundaryScanner.java:397)
	at org.apache.hadoop.hive.ql.udf.ptf.SingleValueBoundaryScanner.computeStart(ValueBoundaryScanner.java:387)
	at org.apache.hadoop.hive.ql.udf.ptf.SingleValueBoundaryScanner.computeStart(ValueBoundaryScanner.java:385)
```

I can put comment for clarity's sake

Reply:
UPDATE: ignore previous comment, this is simply solved by starting from rMin = 0 as per your another comment, thanks!

Reply:
Awesome, thanks Laszlo!